ignite/metrics/ssim.py

```diff
@@ -0,0 +1,171 @@
+        self.sigma = sigma
+        self.data_range = data_range
+        self.k1 = k1
+        self.k2 = k2
+        super(SSIM, self).__init__(output_transform=output_transform)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_ssim = 0.0
+        self._num_examples = 0
+
+    def _gaussian_kernel(self, channel, kernel_size, sigma, device):
+        def gaussian(kernel_size, sigma, device):
+            gauss = torch.arange(
+                start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=torch.float32, device=device
+            )
+            gauss = torch.exp(-gauss.pow(2) / (2 * pow(sigma, 2)))
+            return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+
+        gaussian_kernel_x = gaussian(kernel_size[0], sigma[0], device)
+        gaussian_kernel_y = gaussian(kernel_size[1], sigma[1], device)
+        kernel = torch.matmul(gaussian_kernel_x.t(), gaussian_kernel_y)  # (kernel_size, 1) * (1, kernel_size)
+
+        return kernel.expand(channel, 1, kernel_size[0], kernel_size[1])
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                f"Expected `y_pred` and `y` to have the same data type. Got y_pred: {y_pred.dtype} and y: {y.dtype}."
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                f"Expected `y_pred` and `y` to have the same shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+        if len(y_pred.shape) != 4 or len(y.shape) != 4:
+            raise ValueError(
+                f"Expected `y_pred` and `y` to have BxCxHxW shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+        if self.data_range is None:
+            self.data_range = max(y_pred.max() - y_pred.min(), y.max() - y.min())
+
+        C1 = pow(self.k1 * self.data_range, 2)
+        C2 = pow(self.k2 * self.data_range, 2)
+        device = y_pred.device
+
+        channel = y_pred.size(1)
+        kernel = self._gaussian_kernel(channel, self.kernel_size, self.sigma, device)
+        mu_pred = F.conv2d(y_pred, kernel, groups=channel)
+        mu_target = F.conv2d(y, kernel, groups=channel)
+
+        mu_pred_sq = mu_pred.pow(2)
+        mu_target_sq = mu_target.pow(2)
+        mu_pred_target = mu_pred * mu_target
+
+        sigma_pred_sq = F.conv2d(y_pred * y_pred, kernel, groups=channel) - mu_pred_sq
+        sigma_target_sq = F.conv2d(y * y, kernel, groups=channel) - mu_target_sq
+        sigma_pred_target = F.conv2d(y_pred * y, kernel, groups=channel) - mu_pred_target
+
+        UPPER = 2 * sigma_pred_target + C2
+        LOWER = sigma_pred_sq + sigma_target_sq + C2
+
+        ssim_idx = ((2 * mu_pred_target + C1) * UPPER) / ((mu_pred_sq + mu_target_sq + C1) * LOWER)
+        self._sum_of_batchwise_ssim += torch.mean(ssim_idx, (1, 2, 3))
```

Comment:
Maybe, we have to think about where to store this tensor (CPU or keep original its device)...

Reply:
I would like to keep its original device. What do you think?

Reply:
Yes, let's keep it on original device. Probably, there is no harm as we are not exposing it as a public attribute.